Geofencing/app/src/main/java/com/google/android/gms/location/sample/geofencing/MainActivity.java

```diff
@@ -26,406 +26,411 @@
+  private void performPendingGeofenceTask() {
+    if (mPendingGeofenceTask == PendingGeofenceTask.ADD) {
+      addGeofences();
+    } else if (mPendingGeofenceTask == PendingGeofenceTask.REMOVE) {
+      removeGeofences();
     }
-
-    /**
-     * Performs the geofencing task that was pending until location permission was granted.
-     */
-    private void performPendingGeofenceTask() {
-        if (mPendingGeofenceTask == PendingGeofenceTask.ADD) {
-            addGeofences();
-        } else if (mPendingGeofenceTask == PendingGeofenceTask.REMOVE) {
-            removeGeofences();
-        }
+  }
+
+  /** Return the current state of the permissions needed. */
+  private boolean checkPermissions() {
+    int permissionState =
+        ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION);
+    if (permissionState != PackageManager.PERMISSION_GRANTED) {
+      return false;
     }
-
-    /**
-     * Return the current state of the permissions needed.
-     */
-    private boolean checkPermissions() {
-        int permissionState = ActivityCompat.checkSelfPermission(this,
-                Manifest.permission.ACCESS_FINE_LOCATION);
-        return permissionState == PackageManager.PERMISSION_GRANTED;
+    permissionState =
+        ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_BACKGROUND_LOCATION);
+    return permissionState == PackageManager.PERMISSION_GRANTED;
+  }
+
+  private void requestPermissions() {
+    boolean shouldProvideRationale =
+        ActivityCompat.shouldShowRequestPermissionRationale(
+                this, Manifest.permission.ACCESS_FINE_LOCATION)
+            && ActivityCompat.shouldShowRequestPermissionRationale(
+                this, Manifest.permission.ACCESS_BACKGROUND_LOCATION);
+
+    // Provide an additional rationale to the user. This would happen if the user denied the
+    // request previously, but didn't check the "Don't ask again" checkbox.
+    if (shouldProvideRationale) {
+      Log.i(TAG, "Displaying permission rationale to provide additional context.");
+      showSnackbar(
+          R.string.permission_rationale,
+          android.R.string.ok,
+          view -> {
+            // Request permission
+            ActivityCompat.requestPermissions(
+                MainActivity.this,
+                new String[] {
+                  Manifest.permission.ACCESS_FINE_LOCATION,
+                  Manifest.permission.ACCESS_BACKGROUND_LOCATION
+                },
+                REQUEST_PERMISSIONS_REQUEST_CODE);
+          });
+    } else {
+      Log.i(TAG, "Requesting permission");
+      // Request permission. It's possible this can be auto answered if device policy
+      // sets the permission in a given state or the user denied the permission
+      // previously and checked "Never ask again".
+      ActivityCompat.requestPermissions(
+          MainActivity.this,
+          new String[] {
```

Comment:
This is not allowed since Android 11.

You should first request COARSE + FINE and then BACKGROUND separately

https://developer.android.com/about/versions/11/privacy/location#request-background-location-separately